src/index/vocabulary/SplitVocabulary.h

```diff
@@ -0,0 +1,183 @@
+// Copyright 2025 University of Freiburg
+// Chair of Algorithms and Data Structures
+// Author: Christoph Ullinger <ullingec@cs.uni-freiburg.de>
+
+#ifndef QLEVER_SRC_INDEX_VOCABULARY_SPLITVOCABULARY_H
+#define QLEVER_SRC_INDEX_VOCABULARY_SPLITVOCABULARY_H
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <string_view>
+
+#include "global/ValueId.h"
+#include "index/vocabulary/VocabularyTypes.h"
+#include "util/Exception.h"
+#include "util/HashSet.h"
+
+// Func. to decide where a literal goes
+
+// Func. to get filenames for each vocabulary
+
+// using SplitFunctionT = const std::function<bool(std::string_view)>&;
+// using SplitFilenameFunctionT =
+//     const std::function<std::array<std::string, 2>(std::string)>&;
+
+template <class MainVocabulary, class SpecialVocabulary,
+          const auto& SplitFunction, const auto& SplitFilenameFunction>
+class SplitVocabulary {
+ private:
+  MainVocabulary underlyingMain_;
+  SpecialVocabulary underlyingSpecial_;  // Should not hold conventional string
+                                         // literals (or iris)
+
+  // The 5th highest bit of the vocabulary index is used as a marker to
+  // determine whether the word is stored in the normal vocabulary or the
+  // special vocabulary.
+  static constexpr uint64_t specialVocabMarker = 1ull
+                                                 << (ValueId::numDataBits - 1);
+  static constexpr uint64_t specialVocabIndexMask =
+      ad_utility::bitMaskForLowerBits(ValueId::numDataBits - 1);
+  static constexpr uint64_t maxVocabIndex = specialVocabMarker - 1;
+
+ public:
+  static uint64_t makeSpecialVocabIndex(uint64_t vocabIndex);
+
+  static bool isSpecialVocabIndex(uint64_t index) {
+    return static_cast<bool>(index & specialVocabMarker);
+  }
+
+  static bool isSpecialLiteral(const std::string& input) {
+    return SplitFunction(input);
+  };
+
+  void build(const std::vector<std::string>& words,
+             const std::string& filename);
+
+  void close();
+
+  // Read the vocabulary from files.
+  void readFromFile(const std::string& filename);
+
+  // Problem with getId is that it needs the Comparator which is from the
+  // UnicodeVocab above this splitvocab for the correct lower_bound. -> we can't
+  // do it here
+  // bool getId(std::string_view word, uint64_t* idx) const;
```

Comment:
I don't understand the comment, and the other vocabs don't have a `getId` function.